pkg/trace/agent/agent.go

```diff
@@ -180,59 +183,86 @@ func (a *Agent) loop() {
 // Process is the default work unit that receives a trace, transforms it and
 // passes it downstream.
 func (a *Agent) Process(p *api.Payload) {
-	if len(p.Traces) == 0 {
+	if len(p.Chunks()) == 0 {
 		log.Debugf("Skipping received empty payload")
 		return
 	}
 	defer timing.Since("datadog.trace_agent.internal.process_payload_ms", time.Now())
 	ts := p.Source
-	ss := new(writer.SampledSpans)
+	ss := new(writer.SampledChunks)
 	var envtraces []stats.EnvTrace
 	a.PrioritySampler.CountClientDroppedP0s(p.ClientDroppedP0s)
-	for _, t := range p.Traces {
-		if len(t) == 0 {
+
+	for i := 0; i < len(p.Chunks()); {
+		chunk := p.Chunk(i)
+		if len(chunk.Spans) == 0 {
 			log.Debugf("Skipping received empty trace")
+			p.RemoveChunk(i)
 			continue
 		}
 
-		tracen := int64(len(t))
+		tracen := int64(len(chunk.Spans))
 		atomic.AddInt64(&ts.SpansReceived, tracen)
-		err := normalizeTrace(p.Source, t)
+		err := normalizeTrace(p.Source, chunk.Spans)
 		if err != nil {
 			log.Debugf("Dropping invalid trace: %s", err)
 			atomic.AddInt64(&ts.SpansDropped, tracen)
+			p.RemoveChunk(i)
 			continue
 		}
 
 		// Root span is used to carry some trace-level metadata, such as sampling rate and priority.
-		root := traceutil.GetRoot(t)
+		root := traceutil.GetRoot(chunk.Spans)
+		// Trace chunks from endpoints older than v06 don't have `Priority` in their root.
+		if chunk.Priority == int32(sampler.PriorityNone) && root.Metrics != nil {
+			// if the priority is set in a metric on the root, use it for the whole chunk
+			if p, ok := root.Metrics[tagSamplingPriority]; ok {
+				chunk.Priority = int32(p)
+			}
+		}
+		// Trace chunks from endpoints older than v06 don't have `Origin` in their root.
+		if chunk.Origin == "" && root.Meta != nil {
+			chunk.Origin = root.Meta[tagOrigin]
+		}
+		// Tracer payloads from endpoints older than v06 don't have `Hostname` in their root.
+		if p.TracerPayload.Hostname == "" {
+			if tracerHostname, ok := root.Meta[tagHostname]; ok && tracerHostname != "" {
+				p.TracerPayload.Hostname = tracerHostname
```

Comment:
I'm going to use `v` here. The reason I'm checking if `tagHostname` exists is that we are updating a `TracerPayload` field while iterating the `Chunks` inside `TracerPayload`. If some root spans have `tagHostname` and some don't, we could override `tagHostname` we found with an empty string (`""`).

I can simplify the check with below, since `v` should be `""`, if `ok` is not true.
```
if v := root.Meta[tagHostname]; v != "" {
    ...
}
```

Reply:
Or, I can keep the outer `if` condition, then do as you suggested.

```
if p.TracerPayload.Hostname == "" {
	p.TracerPayload.Hostname = root.Meta[tagHostname]
}
```

Reply:
I'm confused on your first comment. You check if `p.TracerPayload.Hostname` is empty, and only if it is, you assign it the value of the tag. At that point, it does not matter anymore whether the tag is empty or not. 

But I think you already came around on your follow up comment 🙂 So we're good.